components/eamxx/src/diagnostics/pbl_entrainment_budget.cpp

```diff
@@ -0,0 +1,332 @@
+              auto pm_diff = pm_icol(k - 1) - pm_icol(k);
+              if(pblinvalg == 1) {
+                // pblinvalg = 1 ---> based solely on
+                // d(T_mid)/d(p_mid); finding the min (because
+                // d(p_mid) < 0), so keeping signs
+                tm_grad(k) = (tm_icol(k - 1) - tm_icol(k)) / pm_diff;
+              } else if(pblinvalg == 2) {
+                // pblinvalg = 2 ---> based solely on
+                // d(theta_l)/d(p_mid); finding the min
+                // (because d(p_mid) < 0), so keeping signs
+                tm_grad(k) = (tl_icol(k - 1) - tl_icol(k)) / pm_diff;
+              } else if(pblinvalg == 3) {
+                // pblinvalg = 3 ---> based solely on
+                // d(q_t)/d(p_mid); finding the max (because
+                // d(p_mid) < 0), so reversing signs
+                tm_grad(k) = -(qt_icol(k - 1) - qt_icol(k)) / pm_diff;
+              }
+            });
+        team.team_barrier();
+
+        // Find minimum gradient, because d(p_mid) < 0 in definition above
+        // Starting from the surface, and ensuring p_mid > 70000.0 Pa,
+        // to avoid resolving to some odd place higher up in the atmosphere.
+        using minloc_t       = Kokkos::MinLoc<Real, int>;
+        using minloc_value_t = typename minloc_t::value_type;
+        minloc_value_t minloc;
+        Kokkos::parallel_reduce(
+            Kokkos::TeamVectorRange(team, 1, num_levs),
+            [&](const int &k, minloc_value_t &result) {
+              if(tm_grad(k) < result.val && pm_icol(k) > 70000.0) {
+                result.val = tm_grad(k);
+                result.loc = k;
+              }
+            },
+            minloc_t(minloc));
+        team.team_barrier();
+        opt_tm_grad_lev = minloc.loc;
+
+        if(opt_tm_grad_lev < 2 || opt_tm_grad_lev > num_levs - 1) {
+          // Weird stuff can happen near the top and bottom of atm, so fill_val
+          o_pm_hplus(icol) = fill_value;
+          o_tl_hplus(icol) = fill_value;
+          o_qt_hplus(icol) = fill_value;
+          o_df_inpbl(icol) = fill_value;
+          o_tl_caret(icol) = fill_value;
+          o_qt_caret(icol) = fill_value;
+          o_tl_ttend(icol) = fill_value;
+          o_qt_ttend(icol) = fill_value;
+        } else {
+          // Save some outputs just above the "mixed" PBL
+          o_pm_hplus(icol) = pm_icol(opt_tm_grad_lev - 1);
+          o_tl_hplus(icol) = tl_icol(opt_tm_grad_lev - 1);
+          o_qt_hplus(icol) = qt_icol(opt_tm_grad_lev - 1);
+
+          // Save the dF term (F(h) - F(0))
+          o_df_inpbl(icol) = (sd_icol(opt_tm_grad_lev - 1) - sd_icol(0)) -
+                             (su_icol(opt_tm_grad_lev - 1) - su_icol(0)) +
+                             (ld_icol(opt_tm_grad_lev - 1) - ld_icol(0)) -
+                             (lu_icol(opt_tm_grad_lev - 1) - lu_icol(0));
+
+          // Now only need to compute below from opt_tm_grad_lev to num_levs
```

Comment:
Given what discussed above, the integral now starts from the point where tm_grad is the most negative. Is that the idea? Or was it supposed to start from the point where tm_grad changes sign?